contracts/src/NetworkGateway.sol

```diff
@@ -0,0 +1,326 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.25;
+
+import {IFeeVault} from "./payments/interfaces/IFeeVault.sol";
+import {INetworkVerifier} from "./interfaces/INetworkVerifier.sol";
+import {NetworkStorage} from "./NetworkStorage.sol";
+import {NetworkAdmin} from "./NetworkAdmin.sol";
+import {NetworkRegistry} from "./NetworkRegistry.sol";
+
+interface INetworkGatewayEvents {
+    event RequestCallback(
+        uint32 indexed nonce,
+        address indexed verifier,
+        bytes32 programHash,
+        bytes input,
+        bytes context,
+        address callback,
+        bytes4 callbackSelector,
+        uint32 callbackGasLimit,
+        uint256 value
+    );
+    event RequestCall(
+        address indexed verifier,
+        bytes32 programHash,
+        bytes input,
+        address callback,
+        bytes callbackCalldata,
+        uint32 callbackGasLimit,
+        address caller,
+        uint256 value
+    );
+    event FulfilledCallback(
+        uint32 indexed nonce,
+        address indexed verifier,
+        bytes32 programHash,
+        address callback,
+        address prover,
+        bytes32 inputHash,
+        bytes32 outputHash
+    );
+    event FullfilledCall(
+        address indexed verifier,
+        bytes32 programHash,
+        address callback,
+        address prover,
+        bytes32 inputHash,
+        bytes32 outputHash
+    );
+}
+
+interface INetworkGatewayErrors {
+    error ReentrantFulfill();
+    error InvalidRequest(uint32 nonce, bytes32 expected, bytes32 actual);
+    error InvalidProof();
+    error CallbackFailed(bytes4 callbackSelector, bytes output, bytes context);
+    error CallFailed(address callback, bytes callbackData);
+    error InvalidCall(address verifier, bytes input);
+}
+
+interface INetworkGateway is INetworkGatewayEvents, INetworkGatewayErrors {
+    function requestCallback(
```

Comment:
Note: these interfaces wont be backwards compatible with olds ones since verifer/programHash is used instead of functionID